File: 2000_S4.java
//Dynamic programming
import java.util.*;
import java.io.*;
public class Main {
	static StringTokenizer st;
	static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
	static String next () throws IOException {
	  while (st == null || !st.hasMoreTokens())
	  st = new StringTokenizer(br.readLine().trim());
	  return st.nextToken();
	}
	static int readInt () throws IOException {
	  return Integer.parseInt(next());
	}
  
  public static void main(String[] args) throws IOException{ 
    int distance = readInt();
    int num = readInt();
    int[] clubs = new int[num];
    for (int i = 0; i < num; i++) clubs[i] = readInt();
    int[] least = new int[distance + 1]; Arrays.fill(least, 5281);
    least[0] = 0;

    for (int i = 0; i < least.length; i++){
      for (int j = 0; j < clubs.length; j++){
        if (i + clubs[j] <= distance){
          if (least[i] + 1 < least[i + clubs[j]]){
            least[i + clubs[j]] = least[i] + 1;
          }
        }
      }
    }

    if (least[distance] >= 5281) System.out.println("Roberta acknowledges defeat.");
    else System.out.printf("Roberta wins in %d strokes.", least[distance]);
  }
  
}

File: 2000_S4.cpp
//Dynamic programming
#include <bits/stdc++.h>
using namespace std;

typedef long long ll;
const int maxn = 1e5 + 5;

int m, n; //hole length m, n clubs
int clubs[maxn], arr[maxn];

int main() {
  ios_base::sync_with_stdio(false); 
  cin.tie(0); cout.tie(0); 
  cin >> m >> n;
  for (int i = 0; i < n; i++) cin >> clubs[i];
  for (int i = 1; i <= m; i++) arr[i] = 6000;

  for (int i = 1; i <= m; i++){
    for (int j = 0; j < n; j++){
      if (clubs[j] <= i){
        arr[i] = min(arr[i], 1 + arr[i - clubs[j]]);
      }
    }
  }

  if (arr[m] != 6000) 
    cout << "Roberta wins in " << arr[m] << " strokes." << endl;
  else 
    cout << "Roberta acknowledges defeat." << endl;
  
  return 0;
}
